lib/controllers/frontend/spree/checkout_controller_decorator.rb

```diff
@@ -33,11 +33,18 @@ def check_authorization
     # Introduces a registration step whenever the +registration_step+ preference is true.
     def check_registration
       return unless Spree::Auth::Config[:registration_step]
-      return if spree_current_user or current_order.email
+      return if (
+        spree_current_user ||
+        (guest_authenticated? && Spree::Config[:allow_guest_checkout])
```

Comment:
My original thought was that `guest_authenticated?` shouldn't include the configuration check, however, it makes sense that `guest_authenticated?` should always return `false` when `allow_guest_checkout` is configured to `false`.

I went a bit further on the refactoring of this to make it more readable: https://github.com/solidusio/solidus_auth_devise/pull/36/commits/8bc711b5237505c7f854e39a731fc277a50f3353